tests/spiped/test_spiped.sh

```diff
@@ -0,0 +1,254 @@
+fi
+
+# Check for optional commands
+if ! command -v valgrind >/dev/null 2>&1; then
+	echo "valgrind not detected: disabling memory checking"
+	use_valgrind=""
+else
+	use_valgrind="1"
+fi
+
+####################################################
+
+# set up a spiped "encryption" server
+setup_one_spiped () {
+	basename=$1
+
+	# set up valgrind command (if required)
+	if [ -n "$use_valgrind" ]; then
+		valgrind="valgrind --log-file=$out/$basename.val \
+			--leak-check=full --show-leak-kinds=all "
+	else
+		valgrind=""
+	fi
+
+	# start backend server
+	ncat -k -l $dst_port >/dev/null 2>&1 &
+	nc_pid=$!
+	sleep $sleep_ncat_start
+
+	# start spiped to connect source port to backend
+	( $valgrind \
+		$spiped_binary -e \
+	 	-s 127.0.0.1:$src_port -t 127.0.0.1:$dst_port \
+	 	-k keys-blank.txt -F -1 \
+	; echo $? >> $out/$basename.ret ) &
+	sleep $sleep_valgrind_start
+}
+
+# set up spiped "encryption" and "decryption" servers
+setup_two_spipeds () {
+	basename=$1
+
+	# set up valgrind commands (if required)
+	if [ -n "$use_valgrind" ]; then
+		valgrind_e="valgrind --log-file=$out/$basename-e.val \
+			--leak-check=full --show-leak-kinds=all "
+		valgrind_d="valgrind --log-file=$out/$basename-d.val \
+			--leak-check=full --show-leak-kinds=all "
+	else
+		valgrind_e=""
+		valgrind_d=""
+	fi
+
+	# start backend server
+	ncat -k -l -o "$out/$basename.txt" $dst_port >/dev/null 2>&1 &
+	nc_pid=$!
+	sleep $sleep_ncat_start
+
+	# start spiped servers to connect source to mid, and mid to dest ports
+	( $valgrind_e \
+		$spiped_binary -e \
+		-s 127.0.0.1:$src_port -t 127.0.0.1:$mid_port \
+		-k keys-blank.txt -F -1 \
+	; echo $? >> $out/$basename-e.ret ) &
+	( $valgrind_d \
+		$spiped_binary -d \
+		-s 127.0.0.1:$mid_port -t 127.0.0.1:$dst_port \
+		-k keys-blank.txt -F -1 \
+	; echo $? >> $out/$basename-d.ret ) &
+	sleep $sleep_valgrind_start
+}
+
+# set up spiped "encryption" and "decryption" servers, with the "decryption"
+# server being the system
+setup_devel_system_spipeds () {
+	basename=$1
+
+	# set up valgrind commands (if required)
+	if [ -n "$use_valgrind" ]; then
+		valgrind_e="valgrind --log-file=$out/$basename-e.val \
+			--leak-check=full --show-leak-kinds=all "
+		valgrind_d="valgrind --log-file=$out/$basename-d.val \
+			--leak-check=full --show-leak-kinds=all "
+	else
+		valgrind_e=""
+		valgrind_d=""
+	fi
+
+	# start backend server
+	ncat -k -l -o "$out/$basename.txt" $dst_port >/dev/null 2>&1 &
+	nc_pid=$!
+	sleep $sleep_ncat_start
+
+	# start spiped servers to connect source to mid, and mid to dest ports
+	( $valgrind_e \
+		$spiped_binary -e \
+		-s 127.0.0.1:$src_port -t 127.0.0.1:$mid_port \
+		-k keys-blank.txt -F -1 \
+	; echo $? >> $out/$basename-e.ret ) &
+	( $valgrind_d \
+		$system_spiped_binary -d \
+		-s 127.0.0.1:$mid_port -t 127.0.0.1:$dst_port \
+		-k keys-blank.txt -F -1 \
+	; echo $? >> $out/$basename-d.ret ) &
+	sleep $sleep_valgrind_start
+}
+
+
+####################################################
+test_single () {
```

Comment:
Can we have some more useful test names or descriptions?

This test seems to be confirming that if spiped tries to connect to something which doesn't speak the spiped protocol, then spiped will drop the connection?